src/equihash.rs

```diff
@@ -0,0 +1,321 @@
+}
+
+fn tree_validator<'a>(p: &Params, input: &[u8], nonce: &[u8], indices: &[u32]) -> Option<Node> {
+    if indices.len() > 1 {
+        let end = indices.len();
+        let mid = end / 2;
+        match tree_validator(p, input, nonce, &indices[0..mid]) {
+            Some(mut a) => match tree_validator(p, input, nonce, &indices[mid..end]) {
+                Some(mut b) => {
+                    if !has_collision(&a, &b, p.collision_byte_length()) {
+                        error!("Invalid solution: invalid collision length between StepRows");
+                        return None;
+                    }
+                    if b.indices_before(&a) {
+                        error!("Invalid solution: Index tree incorrectly ordered");
+                        return None;
+                    }
+                    if !distinct_indices(&a, &b) {
+                        error!("Invalid solution: duplicate indices");
+                        return None;
+                    }
+                    Some(Node::new(&mut a, &mut b, p.collision_byte_length()))
+                }
+                None => None,
+            },
+            None => None,
+        }
+    } else {
+        let i = indices[0];
+        let mut hash: Vec<u8> = vec![0; p.hash_output()];
+        generate_hash(
+            input,
+            nonce,
+            i / p.indices_per_hash_output(),
+            hash.as_mut_ptr(),
+            p.hash_output(),
+        );
+        let start = ((i % p.indices_per_hash_output()) * p.n / 8) as usize;
+        let end = start + (p.n as usize) / 8;
+        Some(Node {
+            hash: hash[start..end].to_vec(),
+            indices: vec![i],
+        })
+    }
+}
+
+pub fn is_valid_solution_recursive(
+    n: u32,
+    k: u32,
+    input: &[u8],
+    nonce: &[u8],
+    indices: &[u32],
+) -> bool {
+    let p = Params { n: n, k: k };
+
+    match tree_validator(&p, input, nonce, &indices) {
+        Some(root) => {
+            // Hashes were trimmed, so only need to check remaining length
+            root.is_zero(p.collision_byte_length())
+        }
+        None => false,
+    }
+}
+
+pub fn is_valid_solution(n: u32, k: u32, input: &[u8], nonce: &[u8], indices: &[u32]) -> bool {
+    // Recursive validation is faster
+    is_valid_solution_recursive(n, k, input, nonce, indices)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::is_valid_solution_iterative;
+    use super::is_valid_solution_recursive;
+
+    #[test]
+    fn equihash_test_cases() {
+        let input = b"block header";
+        let mut nonce = [0 as u8; 32];
+        let mut indices = vec![
+            976, 126621, 100174, 123328, 38477, 105390, 38834, 90500, 6411, 116489, 51107, 129167,
+            25557, 92292, 38525, 56514, 1110, 98024, 15426, 74455, 3185, 84007, 24328, 36473,
+            17427, 129451, 27556, 119967, 31704, 62448, 110460, 117894,
+        ];
+        assert!(is_valid_solution_iterative(96, 5, input, &nonce, &indices));
```

Comment:
Since you always test against both, make a function that returns the result and panics if the result is different between them.